[TAN-83] Implements attendances/attendees for events

adessy opened the pull request:
# Changelog
- Add support for event registrations (attendance) on the back-end.

adessy committed:
[TAN-83] Add events_attendances table

adessy committed:
[TAN-83] Add unique index on attendee_id-event_id columns in events_attendances table

adessy committed:
[TAN-83] Implement Event::Attendance model + some basic tests

adessy committed:
[TAN-83] Implement a basic endpoint to create event attendances

adessy committed:
[TAN-83] Implement a basic endpoint to delete event attendances

adessy committed:
[TAN-83] Allow admins to unregister any user from events

adessy committed:
[TAN-83] Implement a new endpoint to list all the attendances of an event

adessy committed:
[TAN-83] Remove unnecessary attributes from the attendance serializer

adessy committed:
[TAN-83] add event_attendances association to `User` model

adessy committed:
[TAN-83] Add specs for Events::AttendancePolicy

adessy committed:
[TAN-83] Allow regular users to list their own attendances

adessy committed:
[CL-3966] Add registration info about the current user to serialized events

If the current user is registered for an event, the serialized event will now include a relationship that refers to the corresponding `Events::Attendance` record.

adessy committed:
[CL-3966] Commit some outstanding model annotation from a previous migration

adessy committed:
[TAN-83] Make Event::Finder specs faster

adessy committed:
[TAN-83] Add an endpoint to list all events to which a user is registered

adessy committed:
[TAN-83] Add `attendees` relationship to events

adessy committed:
[TAN-83] Test filtering by `attendee_id` in EventFinder

adessy committed:
[TAN-83] Add side effects for event attendances

adessy committed:
[TAN-83] Fix rubocop offense

adessy committed:
[TAN-83] Fix EventsFinder spec

adessy committed:
[TAN-83] Added a counter cache to events for the number of attendees.

adessy committed:
[TAN-83] Removing debug breakpoint

adessy committed:
[TAN-83] Add attendees_count to serialized events

adessy committed:
[TAN-83] Test that a user can only register once to an event

adessy committed:
[TAN-83] Revert 01ba2d28fb5008b9fd5d29f2f9d839fb6bdca607 partially

Revert "Allow regular users to list their own attendances" (01ba2d28fb5008b9fd5d29f2f9d839fb6bdca607) partially.

Regular users don't need to have access to /events/:id/attendances to check if they are registered for an event. This information is available via:
- /users/:id/events
- the user_attendance attribute of serialized events.

adessy committed:
[TAN-83] Fix rubocop offenses

adessy committed:
[TAN-83] Fix failing test

`with_foreign_key` is actually testing the options passed when creating the association, rather than the existence of the foreign key constraint.

adessy commented on back/app/serializers/web_api/v1/event_serializer.rb:
```diff
@@ -1,11 +1,31 @@
+  attributes(
+    :title_multiloc,
+    :location_multiloc,
+    :attendees_count,
```
⬆️ New attribute here.

adessy commented on back/db/migrate/20230726160134_create_events_attendances.rb:
```diff
@@ -0,0 +1,13 @@
+      t.references :event, null: false, foreign_key: true, type: :uuid, index: true
+
+      t.timestamps
+      t.index %i[attendee_id event_id], unique: true
```
I was thinking about adding an index on timestamps since we order them by `created_at` in the GET .../events/:id/attendances endpoint. However, I then realized that we aren't actually doing this for other models with much more data, such as ideas. What do you think?

kogre replied:
I think it makes sense to add it, given it's not unthinkable that we have many attendances, sometimes.

On ideas, I think it would make sense too, so that's a mistake (at least on published_at)

adessy replied:
OK, I'll add it then.

adessy committed:
[TAN-83] Add some documentation [skip ci]

kogre commented on back/app/controllers/web_api/v1/events_controller.rb:
```diff
@@ -7,7 +7,28 @@ class WebApi::V1::EventsController < ApplicationController
+    # current user is permitted to view the events (since most events are public and
+    # visible to anyone). Instead, we are verifying whether the current user is allowed
+    # to view the events to which the attendee is registered. The events are scoped down
+    # as a second step in `#index` to which this action delegates.
```
To make sure I get this, we're checking whether the user basically has the right to "track" the passed attendee, to know what events they'll be attending? If so, I'd rephrase

> we are verifying whether the current user is allowed
> to view the events to which the attendee is registered

to 

> we are verifying whether the current user has the right to know what the passed attendee is attending

The current phrasing still makes me think more about the events than the attendee

adessy replied:
If it needs explanation, it have to be rephrased :)

kogre commented on back/app/policies/events/attendance_policy.rb:
What about moderators?

For events, from the EventsPolicy seems like we let them manage them, which makes sense, so I assume we should do the same for attendance?

adessy replied:
I was actually thinking to add them in another PR, but maybe I should already add them now.

kogre commented on back/app/serializers/web_api/v1/event_serializer.rb:
```diff
@@ -1,11 +1,32 @@
+    # +:current_user_attendances+ param is used with collections to avoid N+1 queries
+    if (attendances = params[:current_user_attendances])
+      attendances[event.id]
+    elsif signed_in?(event, params)
+      event.attendances.find_by(attendee: params[:current_user])
+    end
```
When are we hitting this branch?

adessy replied:
When serializing a single event. For instance, in:
- `PATCH web_api/v1/events/:id`
- `GET web_api/v1/events/:id`

kogre commented on back/app/serializers/web_api/v1/event_serializer.rb:
```diff
@@ -1,11 +1,32 @@
+    serializer: WebApi::V1::Events::AttendanceSerializer
+  ) do |event, params|
+    # +:current_user_attendances+ param is used with collections to avoid N+1 queries
+    if (attendances = params[:current_user_attendances])
```
This mechanism isn't wrong, per se, just wondering whey we don't this in the more traditional `includes` way. If it's about the parameter, we can add a scope to the model to query for attendances for a specific user and pass that to `includes` or `joins`, I think. Or is there a reason this doesn't work here?

Not blocking, but would be more in line with the conventions if there's no good reason.

adessy replied:
I didn't use `.includes(:attendances)` to avoid instantiating many objects, we don’t actually need and I could not figure out a way to instantiate only the attendance of the current user. If using a scope works, you're my savior :)

By the way, this pattern is employed elsewhere (which is another reason why i thought that was not possible).

adessy replied:
@kogre, how would you do that exactly?

The only solution I could think of looks like this:
```ruby
class Event
  has_one :current_user_attendance, -> { where(attendee: Current.user) }, class_name: 'Events::Attendance'
  # ...
end
```

However, this approach relies on a global state, and `Current.user` does not exist and needs to be implemented.

adessy replied:
As discussed, closing this one because we don't want to rely on global state and parameterized associations are not a thing.

kogre commented on back/app/services/events/side_fx_attendance_service.rb:
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+module Events
+  class SideFxAttendanceService < BaseSideFxService
```
Wasn't aware of BaseSideFxService, not a big fan tbh, but that's another discussion. Added to CDM agenda.

adessy replied:
Can you elaborate, or you prefer keeping this for the CDM? :)

kogre commented on back/config/routes.rb:
```diff
@@ -95,6 +95,8 @@
         resources :comments, only: [:index], controller: 'user_comments'
       end
+
+      get 'users/:attendee_id/events', to: 'events#index_by_attendee'
```
I think I would have nested this under `/events/by_attendee` or something instead, just to have a cleaner colocation between routes and implementation, and more in line with what we've typically done. Not a blocker.

adessy replied:
I don't quite understand what you mean by "cleaner collocation between routes and implementation."

- The use of `by_attendee` seems redundant with `users` (unless we foresee adding an owner to events in the future, but I think that's too much anticipation for an internal API).
- I slightly prefer the name `index_by_attendee` over `by_attendee` for the action name.
- It's not very RESTful ; not that it really matters. However, I usually avoid putting a sub-route where you would expect an identifier when possible.

kogre commented on back/spec/models/event_spec.rb:
```diff
@@ -9,6 +9,16 @@
+  describe 'associations' do
+    subject(:event) { build(:event) }
+
+    it { is_expected.to belong_to(:project) }
+    it { is_expected.to have_many(:attendances).dependent(:destroy) }
+    it { is_expected.to have_many(:attendees).through(:attendances) }
+    it { is_expected.to have_many(:event_files).dependent(:destroy) }
+    it { is_expected.to have_many(:text_images).dependent(:destroy) }
```
Kudos to the extensive coverage, but not too convinced whether these are worth the maintenance time.

adessy replied:
I think they are pretty standard and low cost, including in terms of maintenance. I mean, it took me two minutes to implement. (I'm lying: 3 minutes.)

kogre reviewed (approved):
Clean!

Is the nesting under `Events` module in the root app a pattern we've been using somewhere else already? I get we might want to change things from Engines, but would prefer we do it with a clear aligned direction, instead of introducing new patterns here, and maybe another pattern there.

adessy committed:
Merge branch 'master' into TAN-83/event-improvements/implement-event-attendance

# Conflicts:
#	back/app/models/event.rb
#	back/app/serializers/web_api/v1/event_serializer.rb
#	back/db/schema.rb

adessy committed:
[TAN-83] Improve comment

adessy committed:
[TAN-83] Fix permissions for event attendance

The permissions associated with the use of the `attendee_id` were only verified at the level of EventsController#index_by_attendee. Consequently, those rules were not enforced when calling the `index` action directly with the `attendee_id` parameter.

This commit also adds permission rules for moderators regarding attendances.

adessy committed:
[TAN-83] Adapt Events::AttendancePolicy to take project visibility into account

adessy committed:
[TAN-83] Fix rubocop offense

adessy committed:
[TAN-83] Fix broken test

The test was probably broken when resolving some conflicts with master.
